File: backend/nostr/moderation.py
"""
Moderation — sensitive-content detection (mirror of moderation.js)
------------------------------------------------------------------
Two signals mark a note as adult content:
    1. the author labeled it (NIP-36 "content-warning" tag)
    2. our own keyword screen catches unlabeled text

Returns a human-readable reason, or None if the note looks fine. Nothing is
deleted — the client decides to hide/collapse based on this.
"""

import re

SENSITIVE_WORDS = [
    "nsfw", "porn", "xxx", "nude", "nudes",
    "explicit", "onlyfans", "hentai", "sex", "18+",
]

# Match on word boundaries so "sussex" doesn't trip "sex".
_PATTERN = re.compile(
    r"(^|\W)(" + "|".join(w.replace("+", r"\+") for w in SENSITIVE_WORDS) + r")($|\W)",
    re.IGNORECASE,
)


# Did the author label this note sensitive (NIP-36)?
def content_warning_reason(event):
    tag = next((t for t in event.get("tags", []) if t and t[0] == "content-warning"), None)
    if not tag:
        return None
    return tag[1] if len(tag) > 1 and tag[1] else "sensitive content"


# Full check: author label OR keyword screen. None = looks fine.
def sensitive_reason(event):
    labeled = content_warning_reason(event)
    if labeled:
        return labeled
    if _PATTERN.search(event.get("content", "") or ""):
        return "may contain adult content"
    return None

File: frontend/src/nostr/moderation.js
// ─────────────────────────────────────────────
// Content moderation helpers
// ─────────────────────────────────────────────
// On Nostr nobody can delete content from the network — moderation happens
// in the CLIENT: we decide what to show. Two signals mark a note sensitive:
//
//   1. The author labeled it themselves (NIP-36): a ["content-warning", …]
//      tag, or hashtags like #nsfw. Good clients respect those labels.
//   2. Our own keyword screen catches unlabeled adult text.
//
// Sensitive notes aren't dropped — they collapse behind a "Show" button so
// the user stays in control.

// Words that mark a note as adult content. Matched on word boundaries to
// avoid false positives (e.g. "sussex" must not match "sex").
const SENSITIVE_WORDS = [
  "nsfw",
  "porn",
  "xxx",
  "nude",
  "nudes",
  "explicit",
  "onlyfans",
  "hentai",
  "sex",
  "18+",
];

const SENSITIVE_REGEX = new RegExp(
  `(^|\\W)(${SENSITIVE_WORDS.map((w) => w.replace("+", "\\+")).join("|")})($|\\W)`,
  "i",
);

// Did the author label this note sensitive (NIP-36)?
export function contentWarningReason(event) {
  const tag = event.tags?.find((t) => t[0] === "content-warning");
  return tag ? tag[1] || "sensitive content" : null;
}

// Full check: author label OR our keyword screen.
// Returns a human-readable reason, or null if the note looks fine.
export function sensitiveReason(event) {
  const labeled = contentWarningReason(event);
  if (labeled) return labeled;
  if (SENSITIVE_REGEX.test(event.content || "")) return "may contain adult content";
  return null;
}
